app/src/main/java/ee/ioc/phon/android/speechutils/ContinuousRawAudioRecorder.java

```diff
@@ -0,0 +1,215 @@
+            return mNowPosition;
+        }
+
+        public static SessionStartPointer someMillisBack(int millisBackToStartTheSessionFrom) {
+            return new SessionStartPointer(millisBackToStartTheSessionFrom);
+        }
+
+        public static SessionStartPointer someSecondsBack(int secondsBackToStartTheSessionFrom) {
+            return new SessionStartPointer(secondsBackToStartTheSessionFrom * 1000);
+        }
+    }
+
+    public ContinuousRawAudioRecorder(int audioSource, int sampleRate, int recordingBufferLengthMillis) {
+        super(audioSource, sampleRate, recordingBufferLengthMillis, true);
+
+        // this is very important. We introduce the buffer length to the SessionStartPointer object
+        SessionStartPointer.setRecordingBufferLengthMillis(recordingBufferLengthMillis);
+
+        try {
+            int bufferSize = getBufferSize();
+            int framePeriod = bufferSize / (2 * RESOLUTION_IN_BYTES * CHANNELS);
+            createRecorder(audioSource, sampleRate, bufferSize);
+            createBuffer(framePeriod);
+            setState(State.READY);
+        } catch (Exception e) {
+            if (e.getMessage() == null) {
+                handleError("Unknown error occurred while initializing recorder");
+            } else {
+                handleError(e.getMessage());
+            }
+        }
+    }
+
+    public ContinuousRawAudioRecorder(int sampleRate) {
+        this(DEFAULT_AUDIO_SOURCE, sampleRate, DEFAULT_BUFFER_LENGTH_IN_MILLIS);
+    }
+
+    public ContinuousRawAudioRecorder() {
+        this(DEFAULT_AUDIO_SOURCE, DEFAULT_SAMPLE_RATE, DEFAULT_BUFFER_LENGTH_IN_MILLIS);
+    }
+
+    public String getWsArgs() {
+        return "?content-type=audio/x-raw,+layout=(string)interleaved,+rate=(int)" + getSampleRate() + ",+format=(string)S16LE,+channels=(int)1";
+    }
+
+    public ContinuousRawAudioRecorder setSessionStartPointer(SessionStartPointer sessionStartPointer) {
+        mSessionStartPointer = sessionStartPointer;
+        return this;
+    }
+
+    private int calculateNumOfSamplesToGoBack(int startPos) {
+        // if the consumed session is not the same as the recorded session
+        // get the data from the beginning of the buffer/desired session start pointer
+        if (!isRecordedSessionSameAsConsumedSession()) {
+            Log.i(LOG_FILTER + "Recorded session and consumed session are NOT the same. Grabbing the data from the session start position");
+            return getNumOfSamplesIn(mSessionStartPointer.getSessionStartPointerMillis());
+        }
+
+        Log.i(LOG_FILTER + "Recorded session and consumed session are the same. Working according to the consumed pointer");
+        if (startPos == getLength()) {
+            Log.i(LOG_FILTER + "Consumed session pointer still points to the recorded session pointer. Nothing to do or to return");
+            return -1;
+        }
+
+        int numOfSamplesToGoBack = getLength() - startPos;
```

Comment:
in case that recording pointer is lower than the consume pointer (but didn't pass it yet), the number of samples that should be read are the full buffer minus the gap between the recording pointer and the consume pointer